### DISCUSSION TITLE
[WebDav] Re-Syncing changed files with same filename

### DISCUSSION THREAD

**Discussion by MartB** [photoprism/photoprism#2959] [Q&A] [answered]
https://github.com/photoprism/photoprism/blob/13d32f3294b76c014a4cd6e3ef451f2797126c52/internal/workers/sync_download.go#L104-L110

Can someone tell me if my suspicion is correct that the current webdav implementation will never re-import changed files? It only checks for the download folder but my changed files are not picked up but skipped instead because they already exist.

**Comment by lastzero** [comment]
Either they exist or they were changed.

**Reply by MartB** [reply]
Hmm how do i tell it to not "skip".

It just tells me "skipped" all the time even though the local copy in the originals folder of PhotoPrism that was downloaded is clearly different than the one that's on the WebDav share, the sha1 hashes differ. Do i need to get rid of the temp download folder or bind it to somewhere else?

**Reply by lastzero** [reply]
Reindex the originals.

**Reply by MartB** [reply]
Did not work, i deleted the originals and im waiting for it to trigger a sync now.
Im confused, is this the intended behavior when using a webdav download sync?

Any way to manually trigger the webdav sync?

**Reply by lastzero** [reply]
It compares the hash with what's in the index. If the hash is the same, the content should not be different. That would be like a win in the lottery.

**Reply by MartB** [reply]
Im lost now, it feels like something is broken here but I cant pinpoint what exactly. The Sync shares show up fine and they PROPFIND correctly when selected but somehow nothing gets refreshed. Imma debug this more tomorrow, weird stuff.

**Reply by lastzero** [reply]
Also make sure the modified date changes, MAY depend on the underlying filesystem.

**Reply by MartB** [reply]
Will do :) Thanks for your help so far!

**Comment by MartB** [comment, marked as answer]
Okay just trying to re-cap what i understood so far.

## Refreshing
The refresh worker in sync_refresh.go checks if there were changes in the remote.
For existing files it compares modified dates and if they differ the file status gets changed to `FileSyncNew`. This also happens when a file is completely new.

## Downloading
A second job (sync_download.go) fetches all the entries with `FileSyncNew` state from the photoprism database and downloads them **if they dont exist already.**

Interestingly it only checks for existance and not for the file modification date here. 
> Hence my theory, that if a file was updated it will never get re-fetched again as we instantly stop and set it to `FileSyncExists`.

I do not see how re-syncing of existing files is possible with the current implementation.
If a file with the same name exists when the `os.stat(localName)` call triggers it will not get downloaded no matter if the remote date changed.

Is this assumption correct or am i missing something?
I will modify the code later and play around with updating existing files.

**Reply by lastzero** [reply]
Oh, I understood you were using an app like PhotoSync and the import functionality. It is true that already downloaded files will not be re-downloaded if you use the built-in WebDAV sync feature (where PhotoPrism is the client and not the server), as this would require conflict resolution if the file was also modified in PhotoPrism.

**Reply by MartB** [reply]
The defect description in https://github.com/photoprism/photoprism/issues/552 sounds similar.

**Reply by lastzero** [reply]
Advanced users can use a standard tool like `rsync` for this...

**Reply by lastzero** [reply]
We can add two-way sync with conflict resolution to our paid version once we have funding for a team to develop this :)

**Reply by lastzero** [reply]
If the services are running on the same machine, you can also just mount the same storage folder so you don't even need WebDAV.

**Reply by MartB** [reply]
Yeah sure one can come up with better solutions, just wanted to resort to built-in functionality. Would not hurt to add some modification date check there too i suppose.

Im just leveraging my seafile webdav backend.
Could very well use a davfs2 approach or something similar or a seafile sync client on my machine.

**Reply by MartB** [reply]
> We can add two-way sync with conflict resolution to our paid version once we have funding for a team to develop this :)

To be fair some sort of exclusive "one-way-sync" mode with disabled upload functionality could just re-fetch if something like the modified date is changed. If an user plays around with the originals on the photoprism host its their own problem then.

**Reply by lastzero** [reply]
You could, but what is the consequence then? Re-download everything and then overwrite the local files? Or just download to check the hash and display a warning? Alternatively, you could use a new filename that requires no guesswork and possibly no downloads. Remember that the dates may also change after the files have been migrated to a new server, or because the WebDAV server implementation/time zone has changed. What you are suggesting is easier said than done and could potentially cause a lot of I/O for nothing.

**Reply by MartB** [reply]
For WebDav ETAG comparison could be a feasible solution for some of the points you brought up.
But given that we dont store the ETAG somewhere (its implementation dependent) between servers that would involve a bit of work.

WebDav dates are transmitted with their respective timezones so storing them in UTC should avoid the remote timezone change problem. 

An additional "fix" for the "photoprism is transfered scenario" could be a "modified date worker" that restores the mtime of the files based on the recorded time in the database.

Just saying that there is downsides to everything and i see your points, but the current behavior does not deserve to be called "sync".

PS: I much rather not use webdav but it existed an im a lazy person so theres that haha

**Reply by lastzero** [reply]
I know about all of this. You see, there is only so much a single developer can do. Also, stability is more important to us than shipping lots of features that eventually break or delete / overwrite our users' data.

**Reply by MartB** [reply]
Yeah no doubt about it!
You and all the other contributors to Photoprism did an awesome job, just put that out here for discussions sake 👍
